### ISSUE TITLE
Connect to existing node-process.

### ISSUE THREAD

**Issue by bjackson** [kiddkai/atom-node-debugger#23]
It would be useful to connect to an already running node-process.

**Comment by hfossli** [comment]
Is this possible?

**Comment by kiddkai** [comment]
@hfossli it's possible also doable. I'll try to find out some time to implement it.

**Comment by hfossli** [comment]
Would be awesome! 🎶 

We've embedd our node server in a mac app and starting the node server with NSTask. It would be great to attach this debugger !

**Comment by glachowski** [comment]
:+1

**Comment by maxnowack** [comment]
+1

**Comment by crazyquark** [comment]
+1

**Comment by pcan** [comment]
+1

**Comment by SC7639** [comment]
+1

**Comment by adambuczynski** [comment]
+1

Any chance you could look into this sometime soon? It would be really helpful to debug express apps, as I current cannot seem to do that from within Atom using this plugin.

**Comment by codecontemplator** [comment]
It seems to be one of the mostly requested features so it is definitely a candidate.

**Comment by mhaamann** [comment]
+1

**Comment by strarsis** [comment]
+1

**Comment by inkadnb** [comment]
This would be really useful for Meteor apps too.

**Comment by lpaulger** [comment]
+1

**Comment by jeffmhastings** [comment]
+1

**Comment by gelenbaas** [comment]
+1 !

**Comment by hfossli** [comment]
Any tips to where I should start if I were to implement this?

**Comment by mikidep** [comment]
Is it possible to have it debug a local file but listen to a remotely running instance? i.e. I'd like to debug a project I run in a Vagrant environment.

**Comment by kabo** [comment]
+1

**Comment by pewh** [comment]
+1

**Comment by jschuur** [comment]
+1

**Comment by codyhazelwood** [comment]
+1

**Comment by codecontemplator** [comment]
Thanks for your patience. Made an attempt to implement this in the latest release. Feel free to check it out.

**Comment by adambuczynski** [comment]
Great, thank you. Will definitely try it out.

**Comment by adambuczynski** [comment]
Sorry, but what is the command to attach to an external process? After updating the package in Atom to 1.4.1 the option to attach is no longer present (as described here https://github.com/kiddkai/atom-node-debugger#attaching-to-external-processes)

**Comment by codecontemplator** [comment]
Lets keep this open for feedback.

**Comment by codecontemplator** [comment]
@adambuczynski  I added the attach command to the menu as well now. No keyboard shortcut by default though

**Comment by adambuczynski** [comment]
Thanks, that's fine

**Comment by kmctown** [comment]
Works great! Thanks, this is super useful.

**Comment by adambuczynski** [comment]
This does not seem to work very well for me in my Express app. I run the app using `node --debug=5858 --es_staging --harmony_destructuring scripts/server.js` and then in Atom I try to connect with Node debugger. 

If the app doesn't run, I get the following messages:

```
Connection attempt 3 to node process on localhost:5858 failed. Will try again in 500.
```

This is fine, since the app is not running. Then, after starting the app, the messages stop, and I get a message that Node debugger has connected. So far all good.

But then when I test throwing errors, no output whatsoever shows up in the Node debugger console panel. Nor do I see anything under the call stack etc.

If I introduce a `debugger;` statement in the code, the application breaks on it, but no feedback is shown in Node debugger. However, pressing `resume` does actually step over the `debugger;` breakpoint and continues app execution.

So it seems to work partially, but without output being generated it's not very useful at the moment, at least not for this particular app I'm working on. Any thoughts on what could be going on?

**Comment by yucerjs** [comment]
Hello, Is there a way to map  my project folder to the app path on the server ? 

I would like to debug the code in my working folder, but the debugger downloads a copy for me.

The problem is that my app runs inside a docker container that have my working folder mapped to a volume inside the container that has another path.

I know that the code is exactly the same, but in the server being debugged the code lie in another path.

**Comment by codecontemplator** [comment]
@yucerjs I do not think that is possible. I suggest you open a separate issue for that. Cheers

**Comment by codecontemplator** [comment]
@adambuczynski When the debugger is attached to an external process it will not capture stdout/stderr. That does not explain why the callstack is empty though...

**Comment by adambuczynski** [comment]
@codecontemplator is it not possible for it to capture stdout/stderr?

**Comment by mikidep** [comment]
Trying it out right now, node running in Vagrant VM, if I `attach` a green message gives me some positive feedback, but no reaction whatsoever is given on breakpoints. Also if I try to attach while node is down it just does nothing. No error, no warning, it just opens the widget at the bottom.

**Comment by Khezen** [comment]
@yucerjs

Try to bind src folder from host to src folder from your container. Example with docker-compose:

```
version: '2'
services:

  sample_mongo:
    image: mongo
    expose:
      - "27017"
    volumes:
      - /data/mongo/sample/:/data/db
    network_mode: "bridge"

  sample_node:
    build: ./
    links:
      - sample_mongo:database
    ports:
      - "5858:5858"
    volumes:
      - .:/app
    network_mode: "bridge"
    command: npm debug
``
